teps/0027-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,113 @@
+---
+title: task-exec-status-at-runtime
+authors:
+  - "@pritidesai"
+creation-date: 2020-10-15
+last-updated: 2020-10-19
+status: proposed
+---
+
+# TEP-0027: Task Execution Status at Runtime
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+**Note** In this TEP, `dag` tasks refer to `pipelineTasks` under `tasks` section and `finally` tasks refer to `pipelineTasks` in `finally` section.
+
+A `finally` task in a `pipeline` is executed just before exiting the `pipeline`. `Pipeline` executes `finally` tasks
+after all `dag` tasks finish successfully or one of the `dag` task results in failure i.e `finally` tasks are always
+executed in parallel irrespective of success or failure of a `dag` task. Such `finally` tasks are lacking access to the
+execution status of the `dag` tasks at runtime to better design and process success scenarios v/s failure scenarios.
+
+This proposal is enabling a `finally` task to access the execution status of `dag` tasks at runtime to better design
+`pipeline` based on the status.
+
+## Motivation
+
+* A `pipeline` author wants to send notification to an external system based on the execution status of the `dag` tasks. 
+* A `pipeline` author wants to rollback if the deployment fails in that environment/cluster/region/provider.
+
+### Goals
+
+The main goal of this proposal is to allow access to the execution status of a task within the `pipeline` at runtime.
+And design this approach in the simplest possible way, without any extra processing. 
+
+An additional but important goal of this proposal is identify the different execution states of a `pipelineTask` and
+how each of those states derived based on `taskRun` and/or `pipelineRun` status.
+
+### Non-Goals
+
+* This proposal is not to expose the execution state of `finally` tasks since its not required.
+ 
+* This proposal is not to expose the entire status of a `taskRun`/`pipelineRun` as metadata at runtime.
+
+## Proposal
+
+Introduce a new context variable `$(context.pipelineRun.Tasks.<pipelineTask>.status)` which resolves to one of the execution states:
```

Comment:
> I am wondering if it makes sense to continue following this form for status variables as well?

Following `results` form, would lead us to `$(tasks.<task-name>.status) 🤔 

> Was there a particular reason that this var was nested under context.pipelineRun?

Status is something not generated by the task itself, its something controller generating for each task which I think falls under `context.pipelineRun` similar to pipelineRun name, id, etc. And in future, I am anticipating, this is something auto generated without having task specify explicit mapping i.e. script can directly access anything under $(context.pipelineRun.[*]).

I am happy to change it though.

Reply:
> Status is something not generated by the task itself, its something controller generating for each task which I think falls under context.pipelineRun similar to pipelineRun name, id, etc.

This makes sense, I hadn't thought of it this way, thanks @pritidesai !